worlds/zillion/__init__.py

```diff
@@ -0,0 +1,374 @@
+
+        all: Dict[str, ZillionRegion] = {}
+        for here_zz_name, zz_r in self.zz_system.randomizer.regions.items():
+            here_name = "Menu" if here_zz_name == "start" else zz_reg_name_to_reg_name(here_zz_name)
+            all[here_name] = ZillionRegion(zz_r, here_name, RegionType.Generic, here_name, p, w)
+            self.world.regions.append(all[here_name])
+
+        limited_skill = Req(gun=3, jump=3, skill=self.zz_system.randomizer.options.skill, hp=940, red=1, floppy=126)
+        queue = deque([start])
+        done: Set[str] = set()
+        while len(queue):
+            zz_here = queue.popleft()
+            here_name = "Menu" if zz_here.name == "start" else zz_reg_name_to_reg_name(zz_here.name)
+            if here_name in done:
+                continue
+            here = all[here_name]
+
+            for zz_loc in zz_here.locations:
+                # if local gun reqs didn't place item
+                if not zz_loc.item:
+
+                    def access_rule_wrapped(zz_loc_local: ZzLocation,
+                                            p: int,
+                                            zz_r: ZzRandomizer,
+                                            id_to_zz_item: Dict[int, ZzItem],
+                                            cs: CollectionState) -> bool:
+                        # print(f"checking access to {zz_loc_local}")
+                        accessible = cs_to_zz_locs(cs, p, zz_r, id_to_zz_item)
+                        return zz_loc_local in accessible
+
+                    access_rule = functools.partial(access_rule_wrapped,
+                                                    zz_loc, self.player, self.zz_system.randomizer, self.id_to_zz_item)
+
+                    loc_name = self.zz_system.randomizer.loc_name_2_pretty[zz_loc.name]
+                    loc = ZillionLocation(zz_loc, self.player, loc_name, here)
+                    loc.access_rule = access_rule
+                    if not (limited_skill >= zz_loc.req):
+                        loc.progress_type = LocationProgressType.EXCLUDED
+                        self.world.exclude_locations[p].value.add(loc.name)
+                    here.locations.append(loc)
+                    self.my_locations.append(loc)
+
+            for zz_dest in zz_here.connections.keys():
+                dest_name = "Menu" if zz_dest.name == 'start' else zz_reg_name_to_reg_name(zz_dest.name)
+                dest = all[dest_name]
+                exit = Entrance(p, f"{here_name} to {dest_name}", here)
+                here.exits.append(exit)
+                exit.connect(dest)
+
+                queue.append(zz_dest)
+            done.add(here.name)
+
+    def create_items(self) -> None:
+        if not self.id_to_zz_item:
+            self._make_item_maps("JJ")
+            self.logger.warning("warning: called `create_items` without calling `generate_early` first")
+        assert self.id_to_zz_item, "failed to get item maps"
```

Comment:
if the `if not` really needed?

Reply:
That member will be `None` if `generate_early` hasn't been called.

Not too long ago, Someone made some unit tests that call some functions on the worlds without calling `generate_early` first. When I merged that, the unit tests failed, so I put in safeties just in case these steps aren't all called in the right order.

The `assert` statement after the `if` block is because the type checker can't see that `_make_item_maps` made that member.